Mathlib/AlgebraicGeometry/Grassmannian/Lemmas.lean

```diff
@@ -0,0 +1,245 @@
+They should be included in various other files.
+
+# Notes.
+
+This contribution was created as part of the AIM workshop "Formalizing algebraic geometry" in
+June 2024.
+-/
+
+open AlgebraicGeometry Scheme FiniteDimensional CategoryTheory Matrix
+
+noncomputable section
+
+universe u v w
+
+section
+
+/-- If we give ourselves a scheme gluing data `GD`, and if we have morphisms from the charts
+`GD.U i` of the gluing data to a scheme `Y` that are compatible with the transition morphisms
+between charts, then we get a morphism from the glued scheme `GD.glued` to `Y`. This is an easy
+consequence of the definition of `GD.glued` as a `Multicoequalized`, but is included here for
+convenience.-/
+def AlgebraicGeometry.Scheme.GlueData.glueMorphisms (GD : Scheme.GlueData)
+    {Y : Scheme} (f : (i : GD.J) → GD.U i ⟶ Y) (hf : ∀ (i j : GD.J),
+    GD.f i j ≫ (f i) = GD.t i j ≫ GD.f j i ≫ (f j)) :
+    GD.glued ⟶ Y := by
+  refine Limits.Multicoequalizer.desc _ Y f ?_
+  simp only [GlueData.diagram_l, GlueData.diagram_left, GlueData.diagram_right, Prod.forall,
+    GlueData.diagram_fstFrom, GlueData.diagram_fst, GlueData.diagram_sndFrom, GlueData.diagram_snd,
+    Category.assoc]
+  exact hf
+
+@[simp, reassoc]
+theorem AlgebraicGeometry.Scheme.GlueData.ι_glueMorphisms (GD : Scheme.GlueData) {Y : Scheme}
+    (f : (i : GD.J) → GD.U i ⟶ Y) (hf : ∀ (i j : GD.J), GD.f i j ≫ (f i) =
+    GD.t i j ≫ GD.f j i ≫ (f j)) (i : GD.J) : GD.ι i ≫ GD.glueMorphisms f hf = f i := by
+  erw [Limits.Multicoequalizer.π_desc]
+
+theorem AlgebraicGeometry.Scheme.GlueData.hom_ext (GD : Scheme.GlueData) {Y : Scheme}
+    (f₁ f₂ : GD.glued ⟶ Y) (h : ∀ (i : GD.J), GD.ι i ≫ f₁ = GD.ι i ≫ f₂) : f₁ = f₂ :=
+  GD.openCover.hom_ext f₁ f₂ h
+
+end
+
+section
+
+variable {R A B : Type*} [CommSemiring R] [Semiring A] [Algebra R A] [Semiring B] [Algebra R B]
+
+lemma Algebra.TensorProduct.algebraMap : algebraMap R (TensorProduct R A B) =
+    Algebra.TensorProduct.includeLeftRingHom.comp (algebraMap R A) := by
+  ext _
+  simp only [algebraMap_apply, RingHom.coe_comp, Function.comp_apply, includeLeftRingHom_apply]
+
+end
+
+section
+
+open Function
+
+variable {X Y : Scheme}
+
+lemma SchemeIsoToBijective (f : X ≅ Y) : Bijective f.hom.val.base := by
+  change Bijective ((Scheme.forgetToTop ⋙ forget TopCat).mapIso f).toEquiv
+  exact Equiv.bijective _
+
+end
+
+section
+-- The code in this section was written by Andrew Yang.
+
+open CommRingCat TensorProduct CategoryTheory.Limits
+
+variable (R S T U : Type u) [CommRing R] [CommRing S] [CommRing T] [CommRing U]
+  [Algebra R S] [Algebra R T] [Algebra R U] [Algebra S U] [IsScalarTower R S U] [Algebra T U]
+  [IsScalarTower R T U]
+  (f g : R)
+
+/-- If `S` is an `R`-algebra, this is the morphism of schemes from `Spec S` to `Spec R` defined
+by the structural map from `R` to `S`.-/
+noncomputable
+nonrec abbrev Spec.algebraMap : Spec (.of S) ⟶ Spec (.of R) :=
```

Comment:
Do you think this is useful elsewhere?

Reply:
I don't know. It was in your code so I kept it, but I haven't used it myself.

Reply:
This will disappear once your PR on the pullback of affine schemes is merged, so I'll mark the conversation as resolved.